File: packages/oxlint-plugin-react-doctor/src/plugin/rules/zod/zod-v4-no-deprecated-error-apis.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import {
  getMethodCall,
  getStaticPropertyName,
  getZodNamedImport,
  getZodNamespaceMemberName,
} from "./utils/zod-ast.js";

const DEPRECATED_ZOD_ERROR_MEMBERS = new Set([
  "addIssue",
  "addIssues",
  "errors",
  "flatten",
  "formErrors",
  "format",
]);
const ZOD_ERROR_API_MESSAGE = "Zod 4 dropped this ZodError method, so it breaks when you upgrade.";

const isZodErrorReference = (node: EsTreeNode): boolean => {
  const inner = stripParenExpression(node);
  if (isNodeOfType(inner, "Identifier")) return getZodNamedImport(inner) === "ZodError";
  if (!isNodeOfType(inner, "MemberExpression")) return false;
  return getZodNamespaceMemberName(inner) === "ZodError";
};

const isDirectZodErrorValue = (node: EsTreeNode): boolean => {
  const inner = stripParenExpression(node);
  if (isNodeOfType(inner, "NewExpression")) return isZodErrorReference(inner.callee as EsTreeNode);
  if (!isNodeOfType(inner, "CallExpression")) return false;
  const methodCall = getMethodCall(inner);
  return methodCall?.methodName === "create" && isZodErrorReference(methodCall.receiver);
};

const isDeprecatedZodErrorMemberAccess = (node: EsTreeNode): boolean => {
  if (!isNodeOfType(node, "MemberExpression")) return false;
  const memberExpression = node;
  const propertyName = getStaticPropertyName(memberExpression);
  return (
    propertyName !== null &&
    DEPRECATED_ZOD_ERROR_MEMBERS.has(propertyName) &&
    isDirectZodErrorValue(memberExpression.object as EsTreeNode)
  );
};

const isZodErrorCreateCall = (callExpression: EsTreeNodeOfType<"CallExpression">): boolean => {
  const methodCall = getMethodCall(callExpression);
  return methodCall?.methodName === "create" && isZodErrorReference(methodCall.receiver);
};

const isReceiverOfDeprecatedZodErrorMember = (
  callExpression: EsTreeNodeOfType<"CallExpression">,
): boolean => {
  const parent = callExpression.parent;
  if (!parent || !isNodeOfType(parent, "MemberExpression")) return false;
  if (stripParenExpression(parent.object as EsTreeNode) !== callExpression) return false;
  return isDeprecatedZodErrorMemberAccess(parent);
};

export const zodV4NoDeprecatedErrorApis = defineRule<Rule>({
  id: "zod-v4-no-deprecated-error-apis",
  title: "Deprecated Zod error API",
  requires: ["zod:4"],
  tags: ["migration-hint"],
  severity: "warn",
  recommendation:
    "Use the Zod 4 helpers instead: `z.treeifyError()`, `z.flattenError()`, `z.prettifyError()`, or read `error.issues` directly.",
  create: (context: RuleContext) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      if (isZodErrorCreateCall(node) && isReceiverOfDeprecatedZodErrorMember(node)) return;
      if (!isZodErrorCreateCall(node) && !isDeprecatedZodErrorMemberAccess(node.callee)) {
        return;
      }
      context.report({
        node,
        message: ZOD_ERROR_API_MESSAGE,
      });
    },
    MemberExpression(node: EsTreeNodeOfType<"MemberExpression">) {
      const parent = node.parent;
      if (
        parent &&
        isNodeOfType(parent, "CallExpression") &&
        stripParenExpression(parent.callee as EsTreeNode) === node
      ) {
        return;
      }
      if (!isDeprecatedZodErrorMemberAccess(node)) return;
      context.report({
        node,
        message: ZOD_ERROR_API_MESSAGE,
      });
    },
  }),
});

File: native/oxlint/rules/zod-v4-no-deprecated-error-apis.rs
use oxc_ast::{AstKind, ast::Expression};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{AstNode, context::LintContext, rule::Rule};

const DEPRECATED_ZOD_ERROR_MEMBERS: [&str; 6] = [
    "addIssue",
    "addIssues",
    "errors",
    "flatten",
    "formErrors",
    "format",
];
const MESSAGE: &str =
    "This ZodError API was removed in Zod 4, so error handling can break during the upgrade.";

#[derive(Debug, Default, Clone)]
#[allow(non_camel_case_types)]
pub struct Zod_v4NoDeprecatedErrorApis;

pub type ZodV4NoDeprecatedErrorApis = Zod_v4NoDeprecatedErrorApis;

declare_oxc_lint!(
    /// Warns about Zod 3 error APIs removed in Zod 4.
    Zod_v4NoDeprecatedErrorApis,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Warns about Zod 3 error APIs removed in Zod 4.",
);

impl Rule for Zod_v4NoDeprecatedErrorApis {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let diagnostic_span = match node.kind() {
            AstKind::CallExpression(call_expression) => {
                if is_zod_error_create_call(call_expression, ctx) {
                    if is_receiver_of_deprecated_zod_error_member(node, ctx) {
                        return;
                    }
                } else if !is_deprecated_zod_error_member_expression(
                    &call_expression.callee,
                    ctx,
                ) {
                    return;
                }
                call_expression.span
            }
            AstKind::StaticMemberExpression(member_expression) => {
                if is_call_callee(node, ctx)
                    || !DEPRECATED_ZOD_ERROR_MEMBERS
                        .contains(&member_expression.property.name.as_str())
                    || !is_direct_zod_error_value(&member_expression.object, ctx)
                {
                    return;
                }
                member_expression.span
            }
            AstKind::ComputedMemberExpression(member_expression) => {
                if is_call_callee(node, ctx)
                    || !member_expression.static_property_name().is_some_and(|member_name| {
                        DEPRECATED_ZOD_ERROR_MEMBERS.contains(&member_name.as_str())
                    })
                    || !is_direct_zod_error_value(&member_expression.object, ctx)
                {
                    return;
                }
                member_expression.span
            }
            _ => return,
        };
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(diagnostic_span));
    }
}

fn is_zod_error_reference<'a>(expression: &Expression<'a>, ctx: &LintContext<'a>) -> bool {
    match expression.get_inner_expression() {
        Expression::Identifier(identifier) => direct_named_import_matches(
            identifier,
            &["ZodError"],
            &DIRECT_ZOD_MODULE_SOURCES,
            ctx,
        ),
        expression => expression
            .as_member_expression()
            .is_some_and(|member_expression| {
                member_expression.static_property_name() == Some("ZodError")
                    && matches!(
                        member_expression.object().get_inner_expression(),
                        Expression::Identifier(identifier)
                            if is_direct_zod_namespace_identifier(identifier, ctx)
                    )
            }),
    }
}

fn is_direct_zod_error_value<'a>(expression: &Expression<'a>, ctx: &LintContext<'a>) -> bool {
    match expression.get_inner_expression() {
        Expression::NewExpression(new_expression) => {
            is_zod_error_reference(&new_expression.callee, ctx)
        }
        Expression::CallExpression(call_expression) => {
            is_zod_error_create_call(call_expression, ctx)
        }
        _ => false,
    }
}

fn is_zod_error_create_call<'a>(
    call_expression: &oxc_ast::ast::CallExpression<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    call_expression
        .callee
        .get_inner_expression()
        .as_member_expression()
        .is_some_and(|member_expression| {
            member_expression.static_property_name() == Some("create")
                && is_zod_error_reference(member_expression.object(), ctx)
        })
}

fn is_deprecated_zod_error_member_expression<'a>(
    expression: &Expression<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    expression
        .get_inner_expression()
        .as_member_expression()
        .is_some_and(|member_expression| {
            member_expression.static_property_name().is_some_and(|member_name| {
                DEPRECATED_ZOD_ERROR_MEMBERS.contains(&member_name)
            }) && is_direct_zod_error_value(member_expression.object(), ctx)
        })
}

fn is_receiver_of_deprecated_zod_error_member<'a>(
    call_node: &AstNode<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    let call_root = transparent_expression_root(call_node, ctx);
    let parent = ctx.nodes().parent_node(call_root.id());
    match parent.kind() {
        AstKind::StaticMemberExpression(member_expression) => {
            member_expression.object.get_inner_expression().span() == call_node.span()
                && DEPRECATED_ZOD_ERROR_MEMBERS
                    .contains(&member_expression.property.name.as_str())
        }
        AstKind::ComputedMemberExpression(member_expression) => {
            member_expression.object.get_inner_expression().span() == call_node.span()
                && member_expression.static_property_name().is_some_and(|member_name| {
                    DEPRECATED_ZOD_ERROR_MEMBERS.contains(&member_name.as_str())
                })
        }
        _ => false,
    }
}

fn is_call_callee<'a>(node: &AstNode<'a>, ctx: &LintContext<'a>) -> bool {
    let expression_root = transparent_expression_root(node, ctx);
    matches!(
        ctx.nodes().parent_node(expression_root.id()).kind(),
        AstKind::CallExpression(call_expression)
            if call_expression.callee.get_inner_expression().span() == node.span()
    )
}
